src/node_main_instance.cc

```diff
@@ -65,6 +65,13 @@ NodeMainInstance::~NodeMainInstance() {
   if (isolate_params_ == nullptr) {
     return;
   }
+
+#ifdef DEBUG
+  // node::Environment has been disposed and no JavaScript Execution is allowed
+  // at this point.
+  Isolate::DisallowJavascriptExecutionScope disallow_js(
```

Comment:
Why is `Isolate::DisallowJavascriptExecutionScope` only called when `DEBUG` is set?

Reply:
Disallowing JavaScript execution in this function is primarily a rule to avoid footguns, i.e. `Environment` has been freed and most Node.js APIs would not work. However, it is still valid to evaluate JavaScript as the `Isolate` has not been freed.

So, to enforce the rule, the `Isolate::DisallowJavascriptExecutionScope` is opened and will strictly crash the process if JavaScript is invoked in this scope. This would allow us to identify possible problems in debug build. While in the release build, it would be better to be more lenient and let the JavaScript run at the best effort.